doc: Add UEFI and Windows consumer pages (#1805)
cloud-hypervisor/cloud-hypervisor: docs/windows.md

# Windows Support

> Starting with the release version [0.10.0](https://github.com/cloud-hypervisor/cloud-hypervisor/releases/tag/v0.10.0), Cloud Hypervisor has full Windows guest support.

Author: Applied, thanks.

> Starting with the release version [0.10.0](https://github.com/cloud-hypervisor/cloud-hypervisor/releases/tag/v0.10.0), Cloud Hypervisor has the full Windows guest support.

Reviewer: `Starting with the release version [0.10.0](https://github.com/cloud-hypervisor/cloud-hypervisor/releases/tag/v0.10.0), Cloud Hypervisor has full Windows guest support.`

Requirements

- Host with KVM enabled 
- [UEFI](uefi.md) capable guest

> - Guest with VirtIO integrated drivers

Author: Applied, thanks.

> At the current stage, only UEFI capable Windows images are supported. This implies the presence of the OVMF firmware during the Windows installation an in any subsequent usage. BIOS boot is not supported.

Reviewer: `At the current stage, only UEFI capable Windows images are supported. This implies the presence of the OVMF firmware during the Windows installation and in any subsequent usage. BIOS boot is not supported.`

> The subsequent sections will tell in detail on how to prepare an appropriate Windows image.

Reviewer: `The subsequent sections will tell, in detail, how to prepare an appropriate Windows image.`

Author: Applied your suggestions, thanks for reading through!

## Image Preparation

### Installation using the stock Windows ISO

Prerequisites

- QEMU

> - Windows installation ISO

Author: I googled this https://www.microsoft.com/en-us/evalcenter/evaluate-windows-server-2019 . Another possibility is a Visual Studio subscription. It can be put here, I just didn't see it as relevant.

> - Windows installation ISO

Reviewer 2: Which versions of Windows work? I have only tested Server 2019.

> - Windows installation ISO

Author: I've added a word about this. On my side it is 2019, 2004 and partially 1909.

Thanks.

Parametrizing several command parts, these will be used in the follow up sections as well.

```shell
IMG_FILE=windows-disk.qcow
WIN_ISO_FILE=en_windows_server_version_2004_updated_may_2020_x64_dvd_1e7f1cfa.iso
VIRTIO_ISO_FILE=virtio-win-0.1.185.iso
OVMF_DIR=./FV
```

Create an empty image file, `qcow` or `raw` is supported.
```shell
qemu-img create -f qcow2 $IMG_FILE 30G
```

Begin the Windows installation process under QEMU
```shell
qemu-system-x86_64 -machine q35,accel=kvm \
	-cpu host \
	-m 4G \
	-bios ./$OVMF_DIR/OVMF_CODE.fd \
	-cdrom ./$WIN_ISO_FILE \
	-drive file=./$VIRTIO_ISO_FILE,index=0,media=cdrom
	-drive if=none,id=root,file=./$IMG_FILE \
	-device virtio-blk-pci,drive=root,disable-legacy=on \
	-device virtio-net-pci,netdev=mynet0,disable-legacy=on \
	-netdev user,id=mynet0 \
	-vga std
```

Before the installation can proceed, point the Windows installation program to the VirtIO disk and install the necessary storage controller drivers. After that, the attached hard drive will become visible and the actual installation can commence.

> After the installation has completed, proceed further to the configuration section. QEMU will be needed at least once more to enable SAC.

Author: Applied, thanks.

### Integrating VirtIO drivers into a Windows ISO

> TODO

Reviewer 2: I think you need to give instructions on how to install the network driver

> TODO

Author: There are basically two methods

- Have separate Windows ISO and VirtIO driver ISO. This is what is documented currently.
- Integrate VirtIO stuff into the Windows ISO. Such a prepared Windows ISO will be self sufficient as the VirtIO drivers will be available automatically.

The second method doesn't have to be mentioned, actually But i planned to add some instructions so reserved the space. It all depends on the use case. The second case can be useful for the unattended Windows install.

Regarding the network drivers - once pointed to the driver ISO, Windows will pull them when it comes to the network configuration. It happens automaticallly because the installation program records the path to the ISO and will scan it if it finds some device that is missing drivers.

That's how i saw it happening, no extra steps was needed to install the network driver. If some behavior changes, the procedure of installing drivers is actually always same - the unknown device will be shown in the device manager, one will need the driver iso or exe to install it. 

Thanks.

> TODO

Reviewer 2: I had to manually install the virtio net drivers in the guest (via Device Manager)

> TODO

Author: ACK. I've added now a section wrt device drivers. Underlined the network card specifically.

Thanks.

> TODO

Reviewer 3: If you're planning to detail this section later (follow up PR), then please remove the section as I don't really want to keep `TODO` in it.

## Image Usage

The basic command to boot a Windows image. The configuration section should be checked before executing it for the first time.

```shell
cloud-hypervisor \
	--kernel ./$OVMF_DIR/OVMF.fd \
	--disk path=./$IMG_FILE \
	--cpus boot=1,kvm_hyperv=on \
	--memory size=4G \
	--rng \
	--serial tty \
	--console off \
	--net tap=
```

It is necessary to always

- Carry the OVMF firmware in the `--kernel` option

> - Attach `kvm_hyperv=on` to the `--cpus` option

Author: Applied, thanks.

To daemonize the Cloud Hypervisor process, `nohup` can be used. Some STDIO redirections might need to be done. In a simple case it is sufficient to just redirect all the output to `/dev/null`.

## Image Configuration

> ### SAC 

Reviewer 4: Could we briefly document why we must enable SAC?

Author: Done.

> ### SAC 

Author: Missed this one. Applied, thanks.

Boot the Windows image under QEMU and execute the below commands 

```cmd
bcdedit /emssettings emsport:1 emsbaudrate:115200
bcdedit /ems on
bcdedit /bootems on
```

> Once SAC is enabled, the image can be booted under Cloud Hypervisor. The SAC prompt will show up 

Author: Added a note about the change being permanent. Thanks.

Channel: Cmd0001
SAC>ch -si 1
</pre>

See also the links section for the more extended SAC documentation.

## Network

This section illustrates the Windows specific corner points for the VM network configuration. For the extended networking guide, including bridging for multiple VMs, follow [networking.md](networking.md).

### Basic Networking

As a simplest option, using `--net tap=` in the Cloud Hypervisor command line will create a `vmtapX` device on the host with the default IP4 adress `191.168.249.1`. After SAC becomes available, the guest configuration can be set with

<pre>
SAC>i 10 192.168.249.2 255.255.255.0 192.168.249.1
</pre> 

Where `10` is the device index as output by the `i` command.

### Guest Internet Connectivity

Addinional steps are necessary, to provide the guest with the internet access.

- On the guest, add the DNS server either by using `netsh` or by opening `Network and Connectivity Center` and editing the adapter properties.
- On the host, configure the traffic forwarding
```shell
NET_DEV=wlp3s0
sysctl -w net.ipv4.ip_forward=1
iptables -t nat -A POSTROUTING -o $NET_DEV -j MASQUERADE
```

> ### RDP

Reviewer 4: ```suggestion
### Remote Desktop Protocol (RDP) enablement
```

Author: Applied.

- Using QEMU
  - Execute `SystemPropertiesRemote`
  - In the properties window, choose "Allow remote connections to this computer"
  - Click "Select Users" and add some user to the allow list
- Using powershell
```powershell
   Set-ItemProperty "HKLM:\SYSTEM\CurrentControlSet\Control\Terminal Server\" -Name "fDenyTSConnections" -Value 0
   Enable-NetFirewallRule -DisplayGroup "Remote Desktop"
   Add-LocalGroupMember -Group "Remote Desktop Users" -Member someuser
```
 
Administrators can always RDP, non administrator users have to be explicitly enabled. 

> Once the configuration is set, a favorite RDP client can be used similarly to `rdesktop 192.168.249.2`.

Reviewer 4: ```suggestion
Once the configuration is set, RDP clients can connect to `192.168.249.2`.
```

Author: Applied, thanks.

### SSH

Enable using powershell

```powershell
Add-WindowsCapability -Online -Name OpenSSH.Server~~~~0.0.1.0
Start-Service sshd
Set-Service -Name sshd -StartupType ‘Automatic’
```

> This allows the SSH login from a remote machine the usual way, for example `ssh administrator@192.168.249.2`. For a more detailed OpenSSH guide, please follow the MSDN article from the links section.

Reviewer 4: ```suggestion
This allows for SSH login from a remote machine, for example through the `administrator` user: `ssh administrator@192.168.249.2`. For a more detailed OpenSSH guide, please follow the MSDN article from the [links](#Links) section.
```

Author: Applied, thanks.

# Windows Support

Starting with the release version [0.10.0](https://github.com/cloud-hypervisor/cloud-hypervisor/releases/tag/v0.10.0), Cloud Hypervisor supports Windows guests.

> Requirements

Reviewer 3: I'd suggest `Requirements` to be highlighted by making it bold. `__Requirements__`

Author: Ack, done.

- Host with KVM enabled 
- [UEFI](uefi.md) capable Windows guest image with Virtio drivers integrated

At the current stage, only UEFI capable Windows images are supported. This implies the presence of the OVMF firmware during the Windows installation and in any subsequent usage. BIOS boot is not supported.

The subsequent sections will tell, in detail, how to prepare an appropriate Windows image.

## Image Preparation

### Installation using the stock Windows ISO

> Prerequisites

Reviewer 3: Same here, making `Prerequisites` bold would be nicer. `__Prerequisites__`

```shell
qemu-system-x86_64 -machine q35,accel=kvm \
	-cpu host \
	-m 4G \
	-bios ./$OVMF_DIR/OVMF_CODE.fd \
	-cdrom ./$WIN_ISO_FILE \
	-drive file=./$VIRTIO_ISO_FILE,index=0,media=cdrom
	-drive if=none,id=root,file=./$IMG_FILE \
	-device virtio-blk-pci,drive=root,disable-legacy=on \
	-device virtio-net-pci,netdev=mynet0,disable-legacy=on \
	-netdev user,id=mynet0 \
	-vga std
```

Before the installation can proceed, point the Windows installation program to the VirtIO disk and install the necessary storage controller drivers. After that, the attached hard drive will become visible and the actual installation can commence.

After the installation has completed, proceed further to the configuration section. QEMU will be needed at least once more to enable SAC.

### Integrating VirtIO drivers into a Windows ISO

TODO

## Image Usage

The basic command to boot a Windows image. The configuration section should be checked before executing it for the first time.

```shell
cloud-hypervisor \
	--kernel ./$OVMF_DIR/OVMF.fd \
	--disk path=./$IMG_FILE \
	--cpus boot=1,kvm_hyperv=on \
	--memory size=4G \
	--rng \
	--serial tty \
	--console off \

> 	--seccomp false \

Reviewer 5: Is disabling seccomp still needed? I thought it was fixed.

Author: Yep, just tried without and it works, probably i missed the fix. I'll go remove this part.

Thanks

bcdedit /bootems on
```

Once SAC is enabled, the image can be booted under Cloud Hypervisor. The SAC prompt will show up 

<pre>
Computer is booting, SAC started and initialized.                               
                                                                                
Use the "ch -?" command for information about using channels.                   
Use the "?" command for general help.                                           
                                                                                
                                                                                
SAC>
</pre>

To open a console on the guest, the command sequence below can be used
<pre>
SAC>cmd
The Command Prompt session was successfully launched.
SAC>
EVENT:   A new channel has been created.  Use "ch -?" for channel help.
Channel: Cmd0001
SAC>ch -si 1
</pre>

See also the [links](#Links) section for a more extended SAC documentation.

## Network

This section illustrates the Windows specific corner points for the VM network configuration. For the extended networking guide, including bridging for multiple VMs, follow [networking.md](networking.md).

### Basic Networking

> As a simplest option, using `--net tap=` in the Cloud Hypervisor command line will create a `vmtapX` device on the host with the default IP4 adress `191.168.249.1`. After SAC becomes available, the guest configuration can be set with

Reviewer 2: IP address typo here.

Author: Fixed, thanks.

# Windows Support

Starting with the release version [0.10.0](https://github.com/cloud-hypervisor/cloud-hypervisor/releases/tag/v0.10.0), Cloud Hypervisor supports Windows guests.

__Requirements__

- Host with KVM enabled 
- [UEFI](uefi.md) capable Windows guest image with Virtio drivers integrated

Any modern Windows Server version is compatible. Cloud Hypervisor has been successfully tested with Windows Server 2019 and Windows Server Core 2004.

At the current stage, only UEFI capable Windows images are supported. This implies the presence of the OVMF firmware during the Windows installation and in any subsequent usage. BIOS boot is not supported.

The subsequent sections will tell, in detail, how to prepare an appropriate Windows image.

## Image Preparation

### Installation using the stock Windows ISO

__Prerequisites__

- QEMU
- Windows installation ISO. Obtained through MSDN, Visual Studio subscription, evaluation center, etc. 
- [VirtIO driver ISO](https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/stable-virtio/)
- Suitable [OVMF](uefi.md) firmware 

> - Enough disk space

Reviewer 3: We should provide a minimum size here. This will help our users.

Author: I've put a suggestion of 30G for the image size. Users can deviate OFC.

Thansk.

This step currently requires QEMU to install Windows onto the guest. QEMU is only used at the preparation stage, the resulting image is then fully functional with Cloud Hypervisor.

> Parametrizing several command parts, these will be used in the follow up sections as well.

Reviewer 3: s/Parametrizing/Preparing/
s/, these/as these/

Author: Fixed, thanks.

```shell
qemu-system-x86_64 -machine q35,accel=kvm \
	-cpu host \
	-m 4G \
	-bios ./$OVMF_DIR/OVMF_CODE.fd \
	-cdrom ./$WIN_ISO_FILE \
	-drive file=./$VIRTIO_ISO_FILE,index=0,media=cdrom
	-drive if=none,id=root,file=./$IMG_FILE \
	-device virtio-blk-pci,drive=root,disable-legacy=on \
	-device virtio-net-pci,netdev=mynet0,disable-legacy=on \
	-netdev user,id=mynet0 \
	-vga std
```

Before the installation can proceed, point the Windows installation program to the VirtIO disk and install the necessary storage controller drivers. After that, the attached hard drive will become visible and the actual installation can commence.

After the installation has completed, proceed further to the configuration section. QEMU will be needed at least once more to enable the Windows Special Administration Console (SAC) and to possibly install extra device drivers.

### Integrating VirtIO drivers into a Windows ISO

TODO

## Image Usage

The basic command to boot a Windows image. The configuration section should be checked before executing it for the first time.

```shell
cloud-hypervisor \
	--kernel ./$OVMF_DIR/OVMF.fd \
	--disk path=./$IMG_FILE \
	--cpus boot=1,kvm_hyperv=on \
	--memory size=4G \

> 	--rng \

Reviewer 3: `rng` is enabled by default, so if you're not specifying a different source, there's no need to add it to the command line.

Author: Removed this one, thanks.

	--serial tty \
	--console off \
	--net tap=
```

> It is necessary to always

Reviewer 3: s/always/always:/

Author: Fixed, thanks.

- Carry the OVMF firmware in the `--kernel` option
- Add `kvm_hyperv=on` to the `--cpus` option

To daemonize the Cloud Hypervisor process, `nohup` can be used. Some STDIO redirections might need to be done. In a simple case it is sufficient to just redirect all the output to `/dev/null`.

## Image Configuration

### Device Drivers

> After the Windows installation has finished under QEMU, there might be still devices with no drivers installed. This might happen for example, when a device was not used during the installation. In particular it is important to ensure, that the VirtIO network device is setup correctly, because further steps for the configuration and the usage require network in most case.

Reviewer 3: Please remove the commas as they're not needed.

bcdedit /bootems on
```

Once SAC is enabled, the image can be booted under Cloud Hypervisor. The SAC prompt will show up 

<pre>
Computer is booting, SAC started and initialized.                               
                                                                                
Use the "ch -?" command for information about using channels.                   
Use the "?" command for general help.                                           
                                                                                
                                                                                
SAC>
</pre>

To open a console on the guest, the command sequence below can be used
<pre>
SAC>cmd
The Command Prompt session was successfully launched.
SAC>
EVENT:   A new channel has been created.  Use "ch -?" for channel help.
Channel: Cmd0001
SAC>ch -si 1
</pre>

See also the [links](#Links) section for a more extended SAC documentation.

## Network

This section illustrates the Windows specific corner points for the VM network configuration. For the extended networking guide, including bridging for multiple VMs, follow [networking.md](networking.md).

### Basic Networking

> As a simplest option, using `--net tap=` in the Cloud Hypervisor command line will create a `vmtapX` device on the host with the default IP4 adress `192.168.249.1`. After SAC becomes available, the guest configuration can be set with

Reviewer 3: s/a simplest/the simplest/
s/IP4 adress/IPv4 address/

<pre>
SAC>i 10 192.168.249.2 255.255.255.0 192.168.249.1
</pre> 

> Where `10` is the device index as output by the `i` command.

Reviewer 3: s/output/shown/

### Guest Internet Connectivity

> Addinional steps are necessary, to provide the guest with the internet access.

Reviewer 3: s/Addinional/Additional/
Please remove the comma.
s/with the internet/with internet/

- On the guest, add the DNS server either by using `netsh` or by opening `Network and Connectivity Center` and editing the adapter properties.
- On the host, configure the traffic forwarding
```shell

> NET_DEV=wlp3s0

Reviewer 3: This obviously depends on your host, might be worth mentioning it.

Author: Yep, added a note. Thanks

sysctl -w net.ipv4.ip_forward=1
iptables -t nat -A POSTROUTING -o $NET_DEV -j MASQUERADE
```

### Remote Desktop Protocol (RDP) enablement

> - Using QEMU

Reviewer 3: Use a paragraph for this: `#### Using QEMU`

  - Execute `SystemPropertiesRemote`
  - In the properties window, choose "Allow remote connections to this computer"
  - Click "Select Users" and add some user to the allow list

> - Using powershell

Reviewer 3: Same here: `#### Using Powershell`

Author: Fixed, thanks.

```powershell
   Set-ItemProperty "HKLM:\SYSTEM\CurrentControlSet\Control\Terminal Server\" -Name "fDenyTSConnections" -Value 0
   Enable-NetFirewallRule -DisplayGroup "Remote Desktop"

>    Add-LocalGroupMember -Group "Remote Desktop Users" -Member someuser

Reviewer 3: Remove the spaces at the beginning of each command.

Author: Fixed, thanks.

```
 
Administrators can always RDP, non administrator users have to be explicitly enabled. 

Once the configuration is set, RDP clients can connect to `192.168.249.2`.

### SSH

> Enable using powershell

Reviewer 3: s/Enable using powershell/__Enable using powershell__/

Author: Using teh paragraph as well, thanks.

```powershell
Add-WindowsCapability -Online -Name OpenSSH.Server~~~~0.0.1.0
Start-Service sshd
Set-Service -Name sshd -StartupType ‘Automatic’
```

> This allows for SSH login from a remote machine, for example through the `administrator` user: `ssh administrator@192.168.249.2`. For a more detailed OpenSSH guide, please follow the MSDN article from the [links](#Links) section.

Reviewer 3: s/#Links/#links/

Author: Fixed, thanks.

## Debugging Windows

> TODO

Reviewer 3: Same as the other section with a TODO, if you're planning to add this later, no need to create an empty section with TODO written inside, just remove it and add the section in a follow up PR.

Author: ACK, removing this. Thansk.
